src/main/java/com/powsybl/openloadflow/sensi/AbstractSensitivityAnalysis.java

```diff
@@ -49,13 +47,52 @@ protected AbstractSensitivityAnalysis(MatrixFactory matrixFactory, Supplier<Grap
         this.connectivityProvider = Objects.requireNonNull(connectivityProvider);
     }
 
+    protected static Terminal getEquipmentRegulatingTerminal(Network network, String equipmentId) {
+        Generator generator = network.getGenerator(equipmentId);
+        if (generator != null) {
+            return generator.getRegulatingTerminal();
+        }
+        StaticVarCompensator staticVarCompensator = network.getStaticVarCompensator(equipmentId);
+        if (staticVarCompensator != null) {
+            return staticVarCompensator.getRegulatingTerminal();
+        }
+        TwoWindingsTransformer t2wt = network.getTwoWindingsTransformer(equipmentId);
+        if (t2wt != null) {
+            RatioTapChanger rtc = t2wt.getRatioTapChanger();
+            return rtc != null ? rtc.getRegulationTerminal() : null;
+        }
+        ThreeWindingsTransformer t3wt = network.getThreeWindingsTransformer(equipmentId);
+        Terminal regulatingTerminal = null;
+        if (t3wt != null) {
+            for (ThreeWindingsTransformer.Leg leg : t3wt.getLegs()) {
+                RatioTapChanger rtc = leg.getRatioTapChanger();
+                if (rtc != null && rtc.isRegulating()) {
+                    regulatingTerminal = rtc.getRegulationTerminal();
+                }
+            }
+            if (regulatingTerminal != null) {
+                throw new NotImplementedException(String.format("[%s] Bus voltage on three windings transformer is not managed yet", equipmentId));
+            }
+            return regulatingTerminal;
+        }
+        ShuntCompensator shuntCompensator = network.getShuntCompensator(equipmentId);
+        if (shuntCompensator != null) {
+            return shuntCompensator.getRegulatingTerminal();
+        }
+        VscConverterStation vsc = network.getVscConverterStation(equipmentId);
+        if (vsc != null) {
+            return vsc.getTerminal(); // local regulation only
+        }
+        return null;
+    }
+
     protected JacobianMatrix createJacobianMatrix(EquationSystem equationSystem, VoltageInitializer voltageInitializer) {
         double[] x = equationSystem.createStateVector(voltageInitializer);
         equationSystem.updateEquations(x);
         return new JacobianMatrix(equationSystem, matrixFactory);
     }
 
-    static class LfSensitivityFactor {
+    abstract static class AbstractLfSensitivityFactor {
```

Comment:
Thanks, for this great refactoring! Maybe we can keep a LfSensitivityFactor interface in addition to the AbstractLfSensitivityFactor to avoid using AbstractLfSensitivityFactor everywhere in the code (Abstract class are not supposed to be used as an API)?